### PR TITLE
Move internal domain secret validation to gardener-admission-controller

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity security
/kind cleanup

**What this PR does / why we need it**:
After the introduction of gardenlet, the internal domain secret validation handling was still left to be handled in a better way. With #3753, the gardenlet is restricted to only read `ConfigMap`s related to `Shoot`s for the specific `Seed`s. This doesn't play together nicely with today's `VerifyInternalDomainSecret` function.
Also, #3700 has challenges as well to properly use the `VerifyInternalDomainSecret` function, and we already discussed that it's better to move this into a specific handler in GAC (https://github.com/gardener/gardener/pull/3700#discussion_r593078316).

**Special notes for your reviewer**:
In a future release we have to delete the `garden/gardener-controller-manager-internal-config` `ConfigMap` (it's not used anymore now) - I'll take care to add it for the 1.21 release.

/squash

**Release note**:

```noteworthy operator
The `gardener-admission-controller` does now have a new handler for validating the internal domain `Secret` (earlier, there was no API validation at all).
```

### CODE REVIEW

File: charts/gardener/controlplane/charts/application/templates/validatingwebhook-admission-controller.yaml [hunk 1]
```diff
@@ -109,4 +110,33 @@ webhooks:
...
+    - secrets
+  failurePolicy: Fail
+  clientConfig:
+    {{- if .Values.global.deployment.virtualGarden.enabled }}
+    url: https://gardener-admission-controller.garden/webhooks/admission/validate-internal-domain
+    {{- else }}
+    service:
+      namespace: garden
+      name: gardener-admission-controller
+      path: /webhooks/admission/seedrestriction
```
Comment (BeckerMax) [inline charts/gardener/controlplane/charts/application/templates/validatingwebhook-admission-controller.yaml]:
```suggestion
      path: /webhooks/admission/validate-internal-domain
```

File: hack/local-development/dev-setup-register-gardener [hunk 2]
```diff
@@ -308,7 +310,7 @@ $ADMISSION_CONTROLLER_PORT_STRING
     service:
       namespace: garden
       name: gardener-admission-controller
-      port:
+      port: 2721
```
Comment (BeckerMax) [inline hack/local-development/dev-setup-register-gardener]: why did you add the port specifically? Isn't `$ADMISSION_CONTROLLER_PORT_STRING`  adding the port to the path?

File: hack/local-development/dev-setup-register-gardener [hunk 3]
```diff
@@ -354,6 +356,31 @@ $ADMISSION_CONTROLLER_PORT_STRING
...
+    - UPDATE
+    - DELETE
+    resources:
+    - secrets
+  failurePolicy: Fail
+  clientConfig:
+    service:
+      namespace: garden
+      name: gardener-admission-controller
+      port: 2721
```
Comment (BeckerMax) [inline hack/local-development/dev-setup-register-gardener]: Can we use `$ADMISSION_CONTROLLER_PORT_STRING` here?

File: pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+		secret := &corev1.Secret{}
+		if err := h.decoder.Decode(request, secret); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if _, _, _, _, err := common.GetDomainInfoFromAnnotations(secret.Annotations); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		return acadmission.Allowed("no internal domain secrets exist")
```
Comment (BeckerMax) [inline pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go]:
```suggestion
return acadmission.Allowed("internal domain secret is valid")
```

File: pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+		if exists {
+			return admission.Denied("cannot create internal domain secret because there can be only one secret with the 'internal-domain' secret role")
+		}
+
+		secret := &corev1.Secret{}
+		if err := h.decoder.Decode(request, secret); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if _, _, _, _, err := common.GetDomainInfoFromAnnotations(secret.Annotations); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
```
Comment (BeckerMax) [inline pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go]: `http.StatusUnprocessableEntity` fits better I think

File: pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		_, oldDomain, _, _, err := common.GetDomainInfoFromAnnotations(oldSecret.Annotations)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		_, newDomain, _, _, err := common.GetDomainInfoFromAnnotations(secret.Annotations)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
```
Comment (BeckerMax) [inline pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go]: `http.StatusUnprocessableEntity`

**Review by BeckerMax — CHANGES_REQUESTED** [review]
thanks! have some small requests.

**Review by BeckerMax — DISMISSED** [review]

File: pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	return len(shoots.Items) > 0, nil
+}
+
+func (h *handler) internalDomainSecretExists(ctx context.Context) (bool, error) {
+	secrets := &metav1.PartialObjectMetadataList{}
+	secrets.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("SecretList"))
+
+	if err := h.apiReader.List(
```
Comment (BeckerMax) [inline pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go:158]: As you are listing secrets here do you also need to adapt charts/gardener/controlplane/charts/application/templates/rbac-system.yaml similiar to https://github.com/gardener/gardener/pull/3788 ?

Comment (rfranzke) [reply pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go:158]: Yep, thanks for catching this! Done.

Comment (BeckerMax) [reply pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go:158]: Does a list call always require the watch verb as well?

Comment (rfranzke) [reply pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go:158]: No, but generally, if you allow LIST then WATCH and GET can also be allowed (they are quasi a subset of LIST).

**Review by BeckerMax — APPROVED** [review]
